File: src/blackboard_web/edits.py
"""Local corrections layered over what Blackboard returns.

Blackboard access here is read-only, and a refresh overwrites the cached copy of
every assignment every fifteen minutes. So a correction a student makes — a due
date the instructor moved in class but never in Blackboard, a column name that
means nothing out of context, what each gradebook category is worth — cannot
live in the fetched data. It is kept in its own
cache entry and re-applied on every read, which means a sync can never quietly
undo it.

Nothing here is sent back to Blackboard. These edits change what this dashboard
shows and what its calendar feed publishes, and nothing else.
"""

from __future__ import annotations

from datetime import datetime, timezone
from typing import Any

from .cache import Cache

KEY = "edits"

# What a student is allowed to change about an assignment. Everything else about
# it stays whatever Blackboard says it is.
ASSIGNMENT_FIELDS = ("title", "due_utc", "points_possible", "hidden", "note")


def load(cache: Cache) -> dict[str, Any]:
    stored = cache.get_data(KEY) or {}
    return {
        "assignments": stored.get("assignments") or {},
        "weights": stored.get("weights") or {},
    }


def save(cache: Cache, data: dict[str, Any]) -> dict[str, Any]:
    cache.write(KEY, data)
    return data


def item_key(item: dict[str, Any]) -> str | None:
    """How an assignment is addressed. Matches the key `_track_new` uses."""
    return item.get("column_id") or item.get("content_id")


def set_assignment(cache: Cache, key: str, patch: dict[str, Any]) -> dict[str, Any]:
    """Record an edit. A field set to None is a field handed back to Blackboard."""
    data = load(cache)
    current = dict(data["assignments"].get(key) or {})
    for field in ASSIGNMENT_FIELDS:
        if field not in patch:
            continue
        value = patch[field]
        if value is None or value == "":
            current.pop(field, None)
        else:
            current[field] = value
    if current:
        data["assignments"][key] = current
    else:
        data["assignments"].pop(key, None)
    return save(cache, data)


def clear_assignment(cache: Cache, key: str) -> dict[str, Any]:
    data = load(cache)
    data["assignments"].pop(key, None)
    return save(cache, data)


def set_weights(cache: Cache, course_id: str, weights: dict[str, float]
                ) -> dict[str, Any]:
    """What each gradebook category is worth, in percent, keyed by category id.

    "" addresses the rows the instructor left uncategorised. See
    `sync.category_weights` for how these are read back.
    """
    data = load(cache)
    cleaned = {str(k): float(v) for k, v in (weights or {}).items()}
    if cleaned:
        data["weights"][course_id] = cleaned
    else:
        data["weights"].pop(course_id, None)
    return save(cache, data)


def clear_weights(cache: Cache, course_id: str) -> dict[str, Any]:
    data = load(cache)
    data["weights"].pop(course_id, None)
    return save(cache, data)


def weights_for(cache: Cache, course_id: str) -> dict[str, float]:
    return load(cache)["weights"].get(course_id) or {}


def _reparse(value: str | None) -> datetime | None:
    if not value:
        return None
    try:
        parsed = datetime.fromisoformat(str(value).replace("Z", "+00:00"))
    except ValueError:
        return None
    return parsed if parsed.tzinfo else parsed.replace(tzinfo=timezone.utc)


def apply_assignments(items: list[dict[str, Any]], edits: dict[str, Any],
                      *, keep_hidden: bool = False) -> list[dict[str, Any]]:
    """Return the assignment list as the student has corrected it.

    A moved due date has to carry its derived fields with it — the local time the
    row prints, the countdown beside it and the order of the list all come from
    the due date, so recomputing them here is what makes an edit look real rather
    than half-applied.
    """
    overrides = edits.get("assignments") or {}
    if not overrides:
        return items

    now = datetime.now(timezone.utc)
    out: list[dict[str, Any]] = []
    for item in items:
        key = item_key(item)
        patch = overrides.get(key or "")
        if not patch:
            out.append(item)
            continue
        if patch.get("hidden") and not keep_hidden:
            continue

        edited = {**item, "edited": sorted(k for k in patch if k != "hidden")}
        if patch.get("hidden"):
            edited["hidden"] = True
        if patch.get("title"):
            edited["title"] = patch["title"]
        if patch.get("note"):
            edited["note"] = patch["note"]
        if "points_possible" in patch:
            edited["points_possible"] = patch["points_possible"]
        due = _reparse(patch.get("due_utc"))
        if due:
            edited["due_utc"] = due.isoformat()
            edited["due_local"] = due.astimezone().isoformat()
            edited["days_until"] = round((due - now).total_seconds() / 86400, 1)
        out.append(edited)

    # A moved deadline belongs where its new date puts it, not where its old one
    # left it in the list.
    out.sort(key=lambda a: a.get("due_utc") or "9999")
    return out

File: src/blackboard_web/frontend/src/browser/edits.js
// edits.py for the browser build: local corrections, kept apart from the fetched
// data and re-applied on every read so a sync can never quietly undo one. Nothing
// here reaches Blackboard.

import * as store from "./store.js";
import { cmp, localISO } from "./text.js";

const KEY = "edits";

// What a student may change about an assignment; everything else stays
// whatever Blackboard says it is.
export const FIELDS = ["title", "due_utc", "points_possible", "hidden", "note"];

export async function load() {
  const stored = (await store.getData(KEY)) ?? {};
  return { assignments: stored.assignments ?? {}, weights: stored.weights ?? {} };
}

async function save(data) {
  await store.write(KEY, data);
  return data;
}

/** How an assignment is addressed. Matches the key first-seen tracking uses. */
export const itemKey = (item) => item.column_id || item.content_id || null;

/** Record an edit. A field set to null is a field handed back to Blackboard. */
export async function setAssignment(key, patch) {
  const data = await load();
  const current = { ...(data.assignments[key] ?? {}) };
  for (const field of FIELDS) {
    if (!(field in patch)) continue;
    const value = patch[field];
    if (value === null || value === "") delete current[field];
    else current[field] = value;
  }
  if (Object.keys(current).length) data.assignments[key] = current;
  else delete data.assignments[key];
  return save(data);
}

export async function clearAssignment(key) {
  const data = await load();
  delete data.assignments[key];
  return save(data);
}

/** What each gradebook category is worth, in percent, keyed by category id. */
export async function setWeights(courseId, weights) {
  const data = await load();
  const cleaned = Object.fromEntries(
    Object.entries(weights ?? {}).map(([k, v]) => [String(k), Number(v)]));
  if (Object.keys(cleaned).length) data.weights[courseId] = cleaned;
  else delete data.weights[courseId];
  return save(data);
}

export async function clearWeights(courseId) {
  const data = await load();
  delete data.weights[courseId];
  return save(data);
}

export async function weightsFor(courseId) {
  return (await load()).weights[courseId] ?? {};
}

/** A stored timestamp as a Date. One with no offset is UTC, as in edits.py. */
export function reparse(value) {
  if (!value) return null;
  const s = String(value);
  const d = new Date(/(Z|[+-]\d\d:?\d\d)$/i.test(s) ? s : `${s}Z`);
  return Number.isNaN(d.getTime()) ? null : d;
}

/** The assignment list as the student corrected it.
 *
 * A moved due date carries its derived fields with it — the local time, the
 * countdown and the list order all come from it — or an edit looks half-applied. */
export function applyAssignments(items, edits, { keepHidden = false } = {}) {
  const overrides = edits.assignments ?? {};
  if (!Object.keys(overrides).length) return items;
  const out = [];
  for (const item of items) {
    const patch = overrides[itemKey(item) ?? ""];
    if (!patch) {
      out.push(item);
      continue;
    }
    if (patch.hidden && !keepHidden) continue;
    const edited = { ...item, edited: Object.keys(patch).filter((k) => k !== "hidden").sort() };
    if (patch.hidden) edited.hidden = true;
    if (patch.title) edited.title = patch.title;
    if (patch.note) edited.note = patch.note;
    if ("points_possible" in patch) edited.points_possible = patch.points_possible;
    const due = reparse(patch.due_utc);
    if (due) {
      edited.due_utc = due.toISOString();
      edited.due_local = localISO(due);
      edited.days_until = Math.round((due - Date.now()) / 8640000) / 10;
    }
    out.push(edited);
  }
  // A moved deadline belongs where its new date puts it.
  return out.sort((a, b) => cmp(a.due_utc || "9999", b.due_utc || "9999"));
}
